### PR TITLE
[RISCV] Generalize the (ADD (SLLI X, 32), X) special case in constant…

### PR DESCRIPTION
**Pull request by topperc**
… materialization.

We don't have to limit ourselves to a shift amount of 32. We can support other shift amounts that make the upper 32 bits line up.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -4954,24 +4954,38 @@ static SDValue lowerConstant(SDValue Op, SelectionDAG &DAG,
...
-    int64_t LoVal = SignExtend64<32>(Imm);
-    int64_t HiVal = SignExtend64<32>(((uint64_t)Imm - (uint64_t)LoVal) >> 32);
-    if (LoVal == HiVal ||
-        (Subtarget.hasStdExtZba() && Lo_32(Imm) == Hi_32(Imm))) {
-      RISCVMatInt::InstSeq SeqLo =
-          RISCVMatInt::generateInstSeq(LoVal, Subtarget.getFeatureBits());
-      if ((SeqLo.size() + 2) <= Subtarget.getMaxBuildIntsCost())
-        return Op;
-    }
+  int64_t LoVal = SignExtend64<32>(Imm);
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:4967]: I really dislike the increasing duplication here.

Could we at least introduce a matchShiftedConstant helper?  Have it return the shift amount and instseq for the value by reference?

**Review by preames — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp [hunk 2]
```diff
@@ -195,28 +195,41 @@ static SDValue selectImm(SelectionDAG *CurDAG, const SDLoc &DL, const MVT VT,
...
+      if (MatchedAdd ||
+          (Subtarget.hasStdExtZba() && Lo_32(Imm) == Hi_32(Imm))) {
+        RISCVMatInt::InstSeq SeqLo =
+            RISCVMatInt::generateInstSeq(LoVal, Subtarget.getFeatureBits());
+        if ((SeqLo.size() + 2) < Seq.size()) {
+          SDValue Lo = selectImmSeq(CurDAG, DL, VT, SeqLo);
+
+          SDValue SLLI = SDValue(
+              CurDAG->getMachineNode(RISCV::SLLI, DL, VT, Lo,
+                                     CurDAG->getTargetConstant(32, DL, VT)),
```
[the remark this answers is not in the document]
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:227]: Oops. yes

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 1 above)
Comment (preames) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:4967]: I took a shot at rolling this into the InstSeq abstraction.  I posted it here: https://github.com/llvm/llvm-project/pull/67159

I'm not sold on this approach myself.  I think it's probably only worth doing if we anticipate other cases where we want multiple live values in the same sequence.

**Comment by topperc** [comment]
Moving to draft while I work out issues and refactor.

**Comment by topperc** [comment]
I refactored the duplicate code into a new function in RISCVMatInt.

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp [hunk 3]
```diff
@@ -195,29 +195,24 @@ static SDValue selectImm(SelectionDAG *CurDAG, const SDLoc &DL, const MVT VT,
...
+    RISCVMatInt::InstSeq SeqLo = RISCVMatInt::generateTwoRegInstSeq(
+        Imm, Subtarget.getFeatureBits(), ShiftAmt, AddOpc);
+    if (!SeqLo.empty() && (SeqLo.size() + 2) < Seq.size()) {
+      SDValue Lo = selectImmSeq(CurDAG, DL, VT, SeqLo);
+
+      SDValue SLLI = SDValue(
+          CurDAG->getMachineNode(RISCV::SLLI, DL, VT, Lo,
+                                 CurDAG->getTargetConstant(ShiftAmt, DL, VT)),
+          0);
+      // Prefer ADD when possible.
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:214]: Kill the oddly placed comment

**Review by preames — APPROVED** [review]
LGTM w/minor comment

**Pull request merged**